api/routes/sample.py

```diff
@@ -60,6 +88,46 @@ async def create_new_sample(
         return sample_id_format(internal_id)
 
 
+@router.put(
+    '/{project}/batch', response_model=Dict[str, Any], operation_id='batchUpsertSamples'
+)
+async def batch_upsert_samples(
+    samples: SampleUpsertBatchModel,
+    connection: Connection = get_project_write_connection,
+) -> Dict[str, Any]:
+    """Upserts a list of samples with sequences, and returns the list of internal sample IDs"""
+
+    def upsert_sample(sample):
+        if not sample.id:
+            return create_new_sample(sample, connection)
+        return update_sample(sample.id, sample, connection)
+
+    def upsert_sequence(sid, sequence):
+        sequence.sample_id = sid
+        if not sequence.id:
+            return create_sequence(sequence, connection)
+        return update_sequence(sequence.id, sequence, connection)
+
+    async def upsert_sequences(sid, sequences):
+        upserts = [upsert_sequence(sid, s) for s in sequences]
+        return await asyncio.gather(*upserts)
+
+    async with connection.connection.transaction():
+        # Create or update samples
+        upserts = [upsert_sample(s) for s in samples.samples]
+        runs = await asyncio.gather(*upserts)
+        sids = [sample_id_format(s) for s in runs]
+
+        # Upsert all sequences with paired sids
+        sequences = zip(sids, [x.sequences for x in samples.samples])
+        seqs = [upsert_sequences(sid, seqs) for sid, seqs in sequences]
+        results = await asyncio.gather(*seqs)
```

Comment:
Does this bubble up errors correctly? I can't remember if an exception in a called function causes all other promises to get cancelled.

Reply:
Double checked the error handling. All good using async/await